### PR TITLE
[code-location-concept-page-1] Concept page for code locations [WIP]

### PR DESCRIPTION
**Pull request by schrockn**
Summary & Motivation: This is a WIP, but I wanted to throw up the intro verbiage
to kickstart the conversation here.

The decision I want to discuss here is separating the notion of a python module the has a `Definitions` object defined at module-scope in the `defs` from an entity that is loadable by our tools. This is a subtle but important point. It means that a location doesn't really exist until it it loadable by our tools. So when you type `dagit -f foo.py` it is meant to mean "I am loading this code location in dagit with the current Python environment, communicating with that Python Environment locally, and loading the Python file foo.py as the entry point". This is different than saying that foo.py is a code location, even though undoubtedly it will be referred to it as that colloquially.

Test Plan: Read

### CODE REVIEW

**Comment by schrockn** [comment]
Here is the direct link to the preview. https://dagster-git-schrockn-concept-page-for-code-loca-5931d6-elementl.vercel.app/concepts/code-locations

@yuhan in particular interested in your thoughts on this notion of separating the definition of a "python module entrypoint" from the code location.

File: docs/content/concepts/code-locations.mdx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+# Code Locations
+
+A code location is set of dagster definitions loadable and accessible by Dagster's tools, such as the CLI, Dagit, and Dagster Cloud. A code location comprises:
+
+- A reference to a python module or package that has an instance of `Definitions` at the "defs" variable. We will refer to that as the "entry point".
+- A Python environment that can successfully load the entry point.
+
+Definitions within a code location have a common namespace and must have unique names. This allows them to be grouped and organized by code location in tools.
+
+Before the introduction of the <PyObject object="Definitions" /> API, definitions were grouped into repositories, and there could be many repostories in a particular code location. See the [guide](/concepts/repositories-workspaces/repositories) on Repositories for this previous API and mental model.
```
Comment (yuhan) [inline docs/content/concepts/code-locations.mdx:15]: We can wrap this in a note so it won't get in the way for users who don't care what a repository was

```suggestion
<Note>
Before the introduction of the <PyObject object="Definitions" /> API, definitions were grouped into repositories, and there could be many repostories in a particular code location. See the <a href="/concepts/repositories-workspaces/repositories">guide</a> on Repositories for this previous API and mental model.
</Note>
```

**Review by yuhan — COMMENTED** [review]
I like the summary and like where the page is going. Left a few comments (mostly about formatting and structure).

Besides, should we talk about sequencing here? Shall me or @erinkcochran87 take over this while you're OOO?

File: docs/content/concepts/code-locations.mdx (hunk 1 above)
Comment (schrockn) [reply docs/content/concepts/code-locations.mdx:15]: cool. i'll make this change myself as committing suggestions doesn't play nice with spr

**Comment by gibsondan** [comment]
Separating out the concept of the "code pointer" and the environment where the code is running makes a lot of sense to me. Especially if we can use existing terms for these things rather than introducing new concepts that the user needs to grok, which I think that doc does successfully.

I'm not totally sure the "Python environment" is the right term though? Or at least it doesn't encompass everything that goes into a code location (like the image / container context / etc.) - there's almost yet a third level that defines the container / compute environment in which the Python environment exists. There's also metadata like a git hash / pointer to a git repo.

Like if you look on the Code locations tab in Dogfood there's a bunch of stuff there that's not yet explained on the Concepts page:

image (12)



Running your own grpc server also doesn't fit neatly into this split between code pointer and environment (since you just specify a host and port there, no code pointer required) but I think that's fine.

**Comment by schrockn** [comment]
Yup this is great feedback and I think will be worth a call to talk through. I'm leaning towards pushing out this a week given unfortunate overlapping PTO with key people (you and Sandy) and Community Day this week.

Just to list out the properties of a code location that you list.

* A particular deployment (I'm hypothesizing a consolidation of instance and deployment language here btw) defines how a code location is loaded. 
* A loaded code location has a temporal component (e.g. the commit hash) as well as it's compute environment, as defined by the semantics of the deployment.
* So the language would be "a dagster cloud yaml defines how code locations are packaged into docker container, deployed, and loaded by the dagster cloud infrastructure" 

I think the cost of this language consolidation would be overlap between how people refer to the basket definitions as well as a deployed instance of a code location. The alternative would be introducing a term like "code location server" or something but I would like to avoid that.

**Pull request merged**